src/execution_plan.c

```diff
@@ -733,11 +824,58 @@ static Record* ExecutionPlan_MapNextRecord(ExecutionPlan* ep, ExecutionStep* ste
             record = RG_ErrorRecordCreate(ectx.err, strlen(ectx.err) + 1);
         }
     }
+
+    // asyncRecor should never be returned to us
+    RedisModule_Assert(RedisGears_RecordGetType(record) != asyncRecordType);
+
 end:
 	ADD_DURATION(step->executionDuration);
     return record;
 }
 
+static Record* ExecutionPlan_FilterNextRecord(ExecutionPlan* ep, ExecutionStep* step, RedisModuleCtx* rctx){
+    Record* record = NULL;
+    INIT_TIMER;
+
+    record = ExecutionPlan_NextRecord(ep, step->prev, rctx);
+
+    START_TIMER;
+
+    if(record == NULL){
+        goto end;
+    }
+
+    if(record == &StopRecord || record == &WaitRecord){
+        goto end;
+    }
+
+    if(RedisGears_RecordGetType(record) == errorRecordType){
+        goto end;
+    }
+
+    ExecutionCtx ectx = ExecutionCtx_Initialize(rctx, ep, step);
+    ectx.originRecord = record;
+    int filterRes = step->filter.filter(&ectx, record, step->filter.stepArg.stepArg);
+    if(ectx.err){
+        RedisGears_FreeRecord(record);
+        record = RG_ErrorRecordCreate(ectx.err, strlen(ectx.err) + 1);\
+        filterRes = RedisGears_FilterSuccess; // its not really success but we want to error to continue.
+    }
+
+    if(filterRes == RedisGears_FilterFailed){
+        RedisGears_FreeRecord(record);
+        record = &DummyRecord; // we let the general loop continue of hold;
+        goto end;
+    }
+
+    if(filterRes == RedisGears_FilterHold){
+        record = &DummyRecord; // we let the general loop continue of hold;
+    }
+end:
+    ADD_DURATION(step->executionDuration);
+    return record;
```

Comment:
How will it be async?

Reply:
So why you add 
`RedisModule_Assert(RedisGears_RecordGetType(record) != asyncRecordType);`
line 829 in the map step. 
I asked because I thought the code should be consistent.

Reply:
Because on the map function I might actually get the async record while on filter and foreach we get the status, Pass/Failed/hold.